### ISSUE TITLE
Too many executors

### ISSUE THREAD

**Issue by JosepSampe** [lithops-cloud/lithops#361]
Currently lithops contains one ad-hoc executor for each compute backend in https://github.com/lithops-cloud/lithops/blob/master/lithops/__init__.py, and all executors share the exact same code, what only changes is the compute_backend param. Additionally, as more  compute backends we add, more executors we will have in that file (and more repeated code we will have) which is hard to maintain. Documentation is also hard to maintain and confusing with 10+ executors that have exactly same parameters and code.

I think we should keep the `local_executor()`, and only one generic executor for all the serverless backends called for example `serverless_executor()`, and then if a user wants to switch between backends, he can do it through config.

IMO it is better idea writing applications with a generic executor and then use the config to switch between backends, rather than write an application with ad-hoc executors and then have to modify all the application code if someone wants to run it in another backend.

If someone wants to specify the backend in the code, he can always do: `lithops.serverless_executor(backend='ibm_cf', ...)`

Regarding remote VM execution, I would keep an executor called `standalone_execuor()` (refer to #362)

So, in summary, I suggest to keep 3 executors:
```
lithops.local_executor()
lithops.serverless_executor()
lithops.standalone_executor()
```

**Comment by gilv** [comment]
@JosepSampe function_executor() is generic, right?

**Comment by gilv** [comment]
@erezh16 FYI

**Comment by JosepSampe** [comment]
@gilv function_executor is generic, but to much generic. Now we have to differentiate between serverless and standalone

**Comment by gilv** [comment]
@JosepSampe the idea is great...we also all noticed the "too many" executors that share same code.. Can you please a bit explain in more details the difference between lithops.local_executor() and lithops.standalone_execuor() ?

**Comment by JosepSampe** [comment]
@gilv in fact I think `standalone_executor()` can share some logic of the ` local_executor()` at the end.
The main reason of having these 2 executors is to differentiate in a lithops script what is executed locally, and what is executed remotely. Moreover, `standalone_executor()` has much more parameters than `local_executor(),` for example: cpu, mem, soft_timeout, hard_timeout, auto_dismantle, etc...  Al these parameters are not convenint in the `local_executor()`, so the idea is to separate them.

I think we can also have the `function_executor()` as a top-level executor that will englobe all the rest of executors (local_executor, serverless_executor and standalone_executor). This way, a user has 2 options when writing a lithops application:
1. Create everywhere in the app a `function_executor()`. Then, he will be able to run the whole application locally, in a serverless backend or in a standalone backend by only changing the configuration.
2. Create the appropriate executor in each part of the app: Some times it is well known what kind of executor to use in each part of the application, so a user can force to use local_executor(), serverless_execuor() or standalone_execuor() when convenient.

**Comment by erezh16** [comment]
@JosepSampe Your reasoning seems to be largely the same as what motivated me to propose dropping all executors except the generic one. We can move all the executor specifics into configuration, creating highly simplified and portable Lithops applications. I would further suggest not to have specific executor API at all. If we want to create different specific executors at different stages, we can do so by creating ad-hoc configuration in memory (hard-coded) or explicitly loading new configuration from external locations (soft-coded). The latter approach (soft-coded) should also prove an excellent fit for workflows involving multiple Lithops computations, since it allows separating the various configurations from the application code itself and keeping them in e.g., Kubernetes ConfigMaps.

**Comment by gilv** [comment]
@erezh16 how would it work without executor api?

**Comment by erezh16** [comment]
@gilv That's not what I meant. The generic executor API (map/reduce) **remains**. However, we keep only one type of executor - a generic executor. No need for CF executor, stand-alone executor, docker executor, etc. They all get deprecated.

**Comment by gilv** [comment]
@erezh16 the code of each backend executor is not the same...some uses Flask, some not, some REST calls to CF, some need to clean resources, etc. How can we remove all backends?

**Comment by JosepSampe** [comment]
@gilv we are not talking about to delete the backends, but only to deprecate the high-level executors which in fact all are wrappers of the same `FunctionExecutor` class.

As I stated above, we can keep the `function_executor(`) as a top-level executor, then what @erezh16 suggest would be possible. And also keep (at least at first stage) the `local_executor()`, `serverless_executor()` and `standalone_executor()`, since sometimes in an app, like the EMB pipeline, some parts of the code must be executed serverless and some other parts must be executed serverfull.

**Comment by gilv** [comment]
@JosepSampe @erezh16 i am not sure i fully understand

**Comment by erezh16** [comment]
@gilv I think me and @JosepSampe tend to agree about reducing the number of high-level executors to make the API more stable, portable and sustainable. However, we disagree on what needs to remain. I believe we should only have one generic high-level executor and defer everything else to configuration, to keep the client code (e.g., EMBL applications) completely independent of backend specifics. @JosepSampe believes we should keep a few executors to make them explicitly clear in the code.  

BTW: regarding EMBL. IIRC, the current application probably that sometimes serverless is used and sometimes docker (in VM) is used. However, to experiment with different options, you probably would like the option to experiment with many different configurations without changing the application. Wouldn't it be cleaner and more robust, for example, to assign a **name** to each executor and have the configuration for that name be defined in a global configuration file? OR, as a second option, add a new constructor to the generic executor to create itself out of configuration stored in a given **location**? Then, to easily experiment, you manipulate only the configuration files and run _exactly_ the same application.

**Comment by JosepSampe** [comment]
Resolved, no need to delete backward-compatible executors at the moment
